File: app/src/main/java/com/vpaliy/mediaplayer/media/playback/QueueManager.java
package com.vpaliy.mediaplayer.media.playback;

import android.support.v4.media.MediaMetadataCompat;

import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class QueueManager {

    private List<MediaMetadataCompat> queueData;
    private int index;

    public QueueManager(){
        queueData=new LinkedList<>();
    }

    public void add(MediaMetadataCompat metadata){
        queueData.add(metadata);
    }

    public void setQueueData(List<MediaMetadataCompat> queueData) {
        this.queueData = queueData;
    }

    public MediaMetadataCompat next(){
        if((index+1)<queueData.size()){
            return queueData.get(++index);
        }
        return null;
    }

    public MediaMetadataCompat previous(){
        if(index==0||(index-1)>=queueData.size())
            return null;
        return queueData.get(--index);
    }

    public MediaMetadataCompat current(){
        return !queueData.isEmpty()?queueData.get(index):null;
    }
}

File: app/src/main/java/com/vpaliy/mediaplayer/domain/playback/QueueManager.kt
package com.vpaliy.mediaplayer.domain.playback

import com.vpaliy.mediaplayer.domain.model.Track
import com.vpaliy.mediaplayer.then
import java.util.Collections

class QueueManager(private var tracks: MutableList<Track>, var index: Int) {

  fun next() = !hasNext() then (current()) ?: tracks[++index]

  fun previous() = !hasPrevious() then (current()) ?: tracks[--index]

  fun shuffle() = Collections.shuffle(tracks)

  fun hasNext() = tracks.size > (index + 1)

  private fun hasPrevious() = index - 1 >= 0

  fun size() = tracks.size

  fun addTrack(track: Track) = tracks.add(track)

  fun current() = tracks[index]
}